metals/src/main/scala/scala/meta/internal/metals/debug/BuildTargetClasses.scala

```diff
@@ -271,6 +315,279 @@ final class BuildTargetClasses(val buildTargets: BuildTargets)(implicit
+  ): Unit = {
+    symbolInfo.annotations.foreach { annotation =>
+      annotation.tpe match {
+        case TypeRef(_, annotationSymbol, _) =>
+          TestFrameworkDetector.fromSymbol(annotationSymbol) match {
+            case Some(framework) =>
+              val classSymbol = symbolInfo.symbol
+              val className = symbolToClassName(classSymbol)
+              val testInfo = TestSymbolInfo(className, framework)
+              val classSymbolWithoutFunctionName =
+                classSymbol.indexOf('#') match {
+                  case -1 => classSymbol
+                  case index => classSymbol.substring(0, index + 1)
+                }
+              testClasses += ((classSymbolWithoutFunctionName, testInfo))
+            case None =>
+          }
+        case _ =>
+      }
+    }
+  }
+
+  private def processTestClassHierarchy(
+      symbolInfo: SymbolInformation,
+      doc: TextDocument,
+      path: AbsolutePath,
+      testClasses: scala.collection.mutable.ListBuffer[(String, TestSymbolInfo)],
+  ): Option[Future[Unit]] = {
+    symbolInfo.signature match {
+      case classSig: ClassSignature =>
+        val symbol = symbolInfo.symbol
+        val className = symbolToClassName(symbol)
+        if (className.nonEmpty) {
+          Some(
+            detectTestFrameworkUsingClassHierarchy(classSig, doc, path)
+              .map { frameworkOpt =>
+                frameworkOpt.foreach { framework =>
+                  val testInfo = TestSymbolInfo(className, framework)
+                  testClasses += ((symbol, testInfo))
+                }
+              }
+          )
+        } else None
+      case _ => None
+    }
+  }
+
+  private def detectTestFrameworkUsingClassHierarchy(
+      classSig: ClassSignature,
+      doc: TextDocument,
+      path: AbsolutePath,
+  ): Future[Option[TestFramework]] = {
+    val initialParents = extractParentSymbols(classSig)
+    searchClassHierarchyForTestFramework(
+      initialParents,
+      doc,
+      path,
+    )
+  }
+
+  private def searchClassHierarchyForTestFramework(
+      symbols: List[String],
+      doc: TextDocument,
+      path: AbsolutePath,
+      visited: Set[String] = Set.empty,
+  ): Future[Option[TestFramework]] = {
+    if (symbols.isEmpty) {
+      Future.successful(None)
+    } else {
+      val directFramework = symbols
+        .map(symbol =>
+          TestFrameworkDetector
+            .fromSymbol(symbol)
+        )
+        .collectFirst { case Some(framework) => framework }
+
+      directFramework match {
+        case Some(framework) => Future.successful(Some(framework))
+        case None =>
+          val nextLevelFutures = symbols
+            .filterNot(visited.contains)
+            .map(symbol => collectParentsForSymbol(symbol, doc, path, visited))
+          Future.sequence(nextLevelFutures).flatMap { parentList =>
+            val allParents = parentList.flatten.distinct
+
+            searchClassHierarchyForTestFramework(
```

Comment:
I thought about additionally caching the result of `searchClassHierarchyForTestFramework` for specific symbol. It could potentially speed up computation in cases with big complex test structures, however I don't know if that won't be an overkill. WDYT?

Reply:
I tried to implement some caching of `searchClassHierarchyForTestFramework` but it turned out that it doesn't bring too much value, as in all cases that I checked cache state was invalidated before it was used. I tested something like Base -> Child -> GrandChild and after modifying state of GrandChild `onChange` was called for each document in the chain. 
Caching only `PcSymbolInformation` would be probably sufficient for now